### DISCUSSION TITLE
Is it %{!?foo:...} or %{?!foo:...}

### DISCUSSION THREAD

**Discussion by hroncok** [rpm-software-management/rpm#2340] [Q&A]
At https://rpm-software-management.github.io/rpm/manual/macros.html both variants are used:

>     %{?!macro_name:value}
>     %{!?with_python3:1}

Which syntax is the "proper" one? Or are both of them fully OK?

**Comment by pmatilai** [comment]
Only `%{!?with_python3:1}` actually works.

Thanks for spotting and reporting, I'll fix the manual.

**Reply by pmatilai** [reply]
Erm, scratch that, false testing. Both seem to work, but I only `%{!?with_python3:1}` seems "proper" to me.

**Reply by pmatilai** [reply]
Fixed in https://github.com/rpm-software-management/rpm/commit/7b0b5e7d2cf53817ec5e2545c5f591a26db72cec

The way it's implemented, the order doesn't matter and I've no interest to change it but historically only the `%{!?foo}` syntax has been used and I don't see any reason to recommend otherwise. "not defined" makes more sense than "defined not", unless you're Yoda.

**Reply by hroncok** [reply]
Thanks. See also https://pagure.io/packaging-committee/c/26174464286d7e44c4f5a23f71d487fd9fd4d3f9 for Fedora guidelines.

**Reply by Conan-Kudo** [reply]
Hilariously, `?!` used to not work in debbuild until it was noticed that it worked in rpm, so it was implemented there. 😆
